### DISCUSSION TITLE
File upload

### DISCUSSION THREAD

**Discussion by squishz** [harness-software/wp-graphql-gravity-forms#108] [General]
Hi guys,

Great job on the plugin, been super helpful for a website we are currently building.

I was just wondering, what is the current status off the file upload field? 

As we are gonna need this for a project we are working on, happy to collaborate / work on this depending on the status.

Thanks

**Comment by PudparK** [comment]
I was just about to ask about this. Is there any way I can help?

**Comment by justlevine** [comment]
We hope to have File Uploads done in v0.7 release. 

Currently working on getting  `multipleFiles`  to pass GF validation. Which is requiring _another_ refactor. Which takes a while because there's a lot of technical debt and we're lacking many unit tests.... 

All in good time!

**Comment by justlevine** [comment]
@squishz and @PudparK as of #129, the File Upload and Post Image submissions are now supported. You can learn more about using them [in the docs](https://github.com/harness-software/wp-graphql-gravity-forms/blob/develop/docs/submitting-forms.md)

**Comment by koraysels** [comment]
when I try to uplaod a form with a file i get 'This file type is not permitted for security reasons' 500 error. is this a poblem with wordpress or with graphql-gravityforms?

**Comment by tomwhita** [comment]
WordPress restricts certain file types. This article lists the supported
types:
https://codex.wordpress.org/Uploading_Files#About_Uploading_Files_on_Dashboard

**Reply by koraysels** [reply]
thanks for the reply, I already tried all the formats that are supported including disabling the restriction (with `define('ALLOW_UNFILTERED_UPLOADS', true);
`) but still the error remains..

**Reply by koraysels** [reply]
this is my query:
```gql
  mutation SUBMIT_FORM_ENTRY($formId: Int!, $fieldValues: [FieldValuesInput]) {
    submitGravityFormsForm(
      input: {
        formId: $formId
        clientMutationId: "submitGfForm"
        fieldValues: $fieldValues
      }
    ) {
      clientMutationId
      entryId
      resumeToken
      resumeUrl
      entry {
        form {
          node {
            confirmations {
              message
            }
          }
        }
      }
    }
  }
```

and these are the variables: 

```json
{
    "formId": 2,
    "fieldValues": [
        {
            "id": 5,
            "value": "fweqwef"
        },
        {
            "id": 6,
            "value": "fqfqwf"
        },
        {
            "id": 7,
            "value": "fqwfqwf"
        },
        {
            "id": 8,
            "fileUploadValues": {
                "name": "Rataplan-straatzijde_20180328_EW46474_PHOTO-©-EDDY-WESTVEER-1024x576.jpeg",
                "type": "image/jpeg",
                "size": 162024,
                "tmp_name": "/var/folders/h6/2h8fnrtx1q7ddt18hm162_k80000gn/T/upload_a0051c2799aa83484fc63ae57ffc7a0f"
            }
        }
    ]
}
```

**Reply by justlevine** [reply]
@koraysels To better track and resolve this without spamming the others in this discussion, can you [open a new bug report](https://github.com/harness-software/wp-graphql-gravity-forms/issues/new/choose) ?

that error is being thrown [by our plugin](https://github.com/harness-software/wp-graphql-gravity-forms/blob/develop/src/Utils/GFUtils.php#L472), and is a result of [`wp_check_filtetype_and_ext()`](https://developer.wordpress.org/reference/functions/wp_check_filetype_and_ext/) coming back bad.

FYI: the `ALLOW_UNFILTERED_UPLOADS` only affects administrator roles, but since your example is JPG (which by default is allowed for all users), that shouldnt be the problem here.

Would you be able to add the following code to WP, and then share (in the issue you create) the output generated in debug.log?
```php
add_filter( 
  'wp_check_filetype_and_ext',
  function ( array $file ) { error_log( print_r( $file, true) ); }
);
```

Thanks!

**Reply by koraysels** [reply]
Yes! Thanks a lot I will get to it tomorrow ;)

**Reply by koraysels** [reply]
I opened the following bug report [159](https://github.com/harness-software/wp-graphql-gravity-forms/issues/159#issuecomment-924698679)
